### DISCUSSION TITLE
Support for Multi-Extension FITS images

### DISCUSSION THREAD

**Discussion by ewelot** [libvips/libvips#2080] [Ideas]
The FITS file standard allows for very different arrangement of image data within the file. In the case of storing multiple 2D images of the same size - e.g. from observations of an astronomical target in multiple color bands - there are two typical layouts found:

- A) stack 2D images to form a 3D data cube
- B) put the 2D images in a sequence of consecutive HDUs (Header-Data-Units), often referred to as "extensions"

A special case is the storage of 3 image planes, which could represent a red, green and blue color image. For the variants A and B this will result in

- A) a single extension FITS file, where the primary HDU is an image with NAXIS=3 and NAXIS1=width, NAXIS2=height, NAXIS3=3
- B) a multi-extension FITS file, where each of the 3 HDUs contains an image with NAXIS=2 and NAXIS1=width, NAXIS2=height

As far as I can see from first testing, vips does support the case A - FITS image cubes - only. I'd like to kindly ask to add additional support of the widely used multi-extension FITS image format for reading and writing (the latter might probably need an additional option in vips_fitssave() to distinguish from the default).

Thanks,
Thomas

**Comment by jcupitt** [comment]
Sure, this sounds reasonable. As you say, FITS is a very powerful and flexible format and libvips certainly does not support all of its features.

**Comment by ewelot** [comment]
Much appreciated!
Btw, does the vips image format allows for more than 3 image planes (third dimension >3)?

**Comment by jcupitt** [comment]
Yes, libvips images have three dimensions and all can have any number of elements, up to 2**31.

**Comment by ewelot** [comment]
In this case I would like to suggest following enhancements:

vips_fitsload:
- inspect/load first image HDU in the FITS file
- if it is a 2D image, check/load any other 2D image HDU of the same size
- add option to load a subset of extensions

vips_fitssave:
- write as many image planes as present
- add option to write a subset of image planes
- add option to store as multi-extension FITS instead of a FITS cube

Any specific handling of metadata/FITS headers should be left to the application because there is too many different requirements. It is sufficient if the vips library would just apply a simple conversion: reading and writing header keywords from/to first image HDU (as it is most likely the case already).

Would be awesome to get the multi-extension FITS images supported by vips, as required by my image reduction project. Please let me know if I can help by testing any new implementation.
Thanks, Thomas.

**Comment by ewelot** [comment]
I should provide an example multi-extension FITS file ...
Please find one in the attached zip file. It comprises of the same RGB data as in sky.ppm.
[mef.zip](https://github.com/libvips/libvips/files/5580397/mef.zip)

**Comment by jcupitt** [comment]
Thanks for this -- let's leave this open as an enhancement issue.
